### ISSUE TITLE
Keyword continue is ignored in some cases

### ISSUE THREAD

**Issue by sergutsan** [SophieKoonin/JavaDecaf#17]
**Description of the problem:**
In some cases, the keyword `continue` is ignored by the Java Decaf compiler, for example in the following code: 

```
int n = 0;
while (n < 10) {
   n++;
  if (n % 2 == 0) {
    continue; // this continue is not ignored but it is 
  }                // indented weirdly (is this a clue?)
  println(n);
}
while (n < 20) {
  n++;
  if (n % 2 == 0) {
    println("Skipping even number");
    continue; // this continue is ignored!
  }
  println(n);
}
```

**What should you see?**
Something like 

```
(...)
    int n = 0;
    while (n < 10) {
        n++;
        if (n % 2 == 0) {
            continue;
        }
        System.out.println(n);
    }
    while (n < 20) {
        n++;
        if (n % 2 == 0) {
            System.out.println("Skipping even number");
            continue;
        }
        System.out.println(n);
    }
}
```

**What do you see instead?**

```
(...)
    int n = 0;
    while (n < 10) {
        n++;
        if (n % 2 == 0) {
continue;
        }
        System.out.println(n);
    }
    while (n < 20) {
        n++;
        if (n % 2 == 0) {
            System.out.println("Skipping even number");
        }
        System.out.println(n);
    }
}
```

**Comment by sophiekoonin** [comment]
The grammar needs editing to introduce a new node #ContinueStatement. 

Current:

```
void ContinueStatement() :
{}
{
  "continue" [ <IDENTIFIER> ] ";"
}
```

Suggested:

```
void ContinueStatement() :
{Token t;}
{ ({t = getToken(1);}
  "continue" [ <IDENTIFIER> ] ";"
  {setFirstLastToken(jjtThis,t,getToken(0));})#ContinueStatement
}
```

Upon rebuilding the grammar this will create a class ASTContinueStatement which will need to be edited to implement Indentable. This should fix it...

**Comment by sergutsan** [comment]
Trivial question: how is the grammar rebuilt?

**Comment by sophiekoonin** [comment]
By running Rebuild.sh in the javacc folder. You need JavaCC installed and
configured with path variables.

It's a bit more tedious on Windows, as I couldn't get a batch file working
to do the same, so you have to run it manually:

```
jjtree JavaDecaf.jjt
```

then move the generated JavaDecaf.jj file from the AST folder to the javacc
folder (couldn't configure where it saves the jj file!)
then run 

```
javacc JavaDecaf.jj
```

**Comment by sergutsan** [comment]
...and this (I assume) creates new Java files for the new nodes but does not overwrite those nodes already in place, right?

Thanks a lot in advance.

**Comment by sophiekoonin** [comment]
Exactly!

**Comment by sergutsan** [comment]
I see a message 

```
Warning: TokenMgrError.java: File is obsolete.  Please rename or delete this file 
                  so that a new one can be generated for you.
Warning: ParseException.java: File is obsolete.  Please rename or delete this file
                  so that a new one can be generated for you.
```

It would appear that JavaCC does intend to overwrite some files but is well-behaved and does not do it if you have edited them at some point. 

What I find surprising is that it only tries to overwrite _some_ files, not all (e.g. `JDCParser`).

**Comment by sergutsan** [comment]
First attempt in af6a641c6453f6e8c463011c7c88f52ce85f0691. 
Good news: we get all the continues. :-)
Bad news: we lose all the semicolons. :-(
Will look at this asap: may not be that difficult to fix (_famous last words_). Now I must leave this building.

**Comment by sophiekoonin** [comment]
Those warnings can be ignored. Go home!!

**Comment by sergutsan** [comment]
I said before: "_It would appear that JavaCC does intend to overwrite some files but is well-behaved and does not do it if you have edited them at some point. What I find surprising is that it only tries to overwrite some files, not all (e.g. JDCParser)_". 

Actually, it _does overwrite_ JDCParser. It is not well-behaved at all. When we created ASTContinueStatement, we lost all the latest changes to JDCParser (e.g. d6221c59c24b2c7, df422ec5e548b3b, 01e8ee9a37fab1ce681). :-(

I realised out of chance. Issue #7 should rise in priority so that we do not get these regressions anymore ---this should have been noted before merging to master. Fortunately enough, at least we keep all the code thanks to git. I will recover the changes manually for now.

**Comment by sergutsan** [comment]
Or maybe these changed to JDCParser should actually be written somewhere else and not directly on JDCParser?

**Comment by sophiekoonin** [comment]
JDCParser gets regenerated every time so it's a bad idea to edit it! The
code you want to put in should go at the top of the JavaDecaf.jjt grammar
(you will see the other methods there) and that will appear in JDCParser
when you rebuild.

**Comment by sergutsan** [comment]
Thanks, Sophie. I was imagining something like that. Will fix asap (working on Issue #22 now).

**Comment by sergutsan** [comment]
Done on d1c3d7971b24e831e0ce8d4fa714ac01ab7176ad. 

I guess this closes this issue.
